submariner/cable-driver-policy.md

```diff
@@ -0,0 +1,250 @@
+
+## Proposal
+
+This enhancement proposes a new policy to allow users to selectively deploy different cable types between clusters
+in a ClusterSet. The proposal assumes that connection between two clusters is symmetrical. If Cluster A is connected
+to Cluster B via a VXLAN cable type - then Cluster B is connected to Cluster A via a VXLAN cable type and not another
+cable type.
+
+This enhancement will propose a single API for cable driver policy support using custom resource definition (CRDs).
+
+### Cable driver policy API
+
+This following CRD is proposed:
+
+```Go
+type CablePolicies struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    ConnectionPolicies []CablePolicySpec `json:"spec"`
+}
+
+type CablePolicySpec struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    // Name of the cable driver implementation to use for this connection.
+    CableDriver string `json:"name"`
+
+    // Options specifies a ConfigMap to use for additional configuration parameters for the cable driver.
+    CableOptions *v1.ConfigMap `json:"options,omitempty"`
+}
+```
+
+The `CablePolicies` CR defines the criteria for selecting which cable driver to use to connect cluster pairs (or subsets)
+and their associated (optional) configuration options. To resolve the connectivity policy for a remote cluster's Endpoint,
+Submariner would:
+
+- Search for a policy whose left and right label selectors match the corresponding Cluster resources of the local
+  and remote Endpoints.
+
+If any matching policy is found, the clusters are connected using the cable driver specified in the cable driver policy.
+
+If no matching policy is found, the `default` cable driver policy is used. The `default` policy can be auto-created at broker
+deploy time.
+
+> **_NOTE:_** In the case of multiple labels being specified, the selector will need to match all the labels.
+> Equality-based requirements will be supported. For e.g. labels such as `env: !production` will match any clusters
+> that aren't labelled with `env: production`.
+
+The initial `CablePolicies` resources would be created (at startup time) and maintained on the broker cluster
+and synced to each participating cluster. Policies can subsequently be created/deleted on a cluster and will be
+automatically exported to the broker. A controller to reconcile updates to policies will need to be implemented in
+Submariner.
```

Comment:
Why do we want to allow updates on participating clusters? Presumably the clusterset administrator has access to the broker.